src/mo_types/cons.ml

```diff
@@ -41,11 +51,38 @@ let unsafe_set_kind c k = c.kind := k
 let name c = c.name
 
 let to_string show_stamps sep c =
-  if not show_stamps || c.stamp = 0 then c.name else Printf.sprintf "%s%s%i" c.name sep c.stamp
+  (* Filepaths may have non-parseable characters, so we convert them to hex. *)
+  let escape_char c =
+    match c with
+    | 'a'..'z' | 'A'..'Z' | '0'..'9' | '_' -> String.make 1 c
+    | _ -> Printf.sprintf "%02X" (Char.code c)
+  in
+  let escape_filepath f =
+    let n = String.length f in
+    let buffer = Buffer.create (n * 2) in
+    for i = 0 to n - 1 do
+      Buffer.add_string buffer (escape_char f.[i])
+    done;
+    Buffer.contents buffer
+  in
+  if not show_stamps
+  then c.name
+  else
+    Printf.sprintf
+      "%s%s%i%s"
+      c.name
+      sep
+      (fst c.stamp)
+      (match snd c.stamp with
+      | None -> ""
+      | Some scope -> sep ^ escape_filepath scope)
 
 let eq c1 c2 = c1.stamp = c2.stamp && c1.name = c2.name
 
 let compare c1 c2 =
-  match Int.compare c1.stamp c2.stamp with
-  | 0 -> String.compare c1.name c2.name
+  match Int.compare (fst c1.stamp) (fst c2.stamp) with
+  | 0 ->
+    (match Option.compare String.compare (snd c1.stamp) (snd c2.stamp) with
```

Comment:
I would first compare the int, then the strings, for better perf. Type checking is actually a bottleneck.

Reply:
So now the order is lexicographic order of tuples `(stamp_number, stamp_scope, name)` .
Which order do you think makes more sense?

Reply:
I have the same understanding as Pavel, since the numbers and names are more likely to be different than the scopes, then perhaps `(stamp_number, name, stamp_scope)` would make more sense. Should we proceed with this?

Reply:
Yes, I think so. (stamp_number, name, stamp_scope)